File: projects/android-readplace-poc/app/src/main/kotlin/com/readplace/poc/platform/HtmlCaptor.kt
package com.readplace.poc.platform

import android.annotation.SuppressLint
import android.content.Context
import android.webkit.WebResourceError
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import com.readplace.poc.core.AppConfig
import kotlinx.coroutines.delay
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withTimeoutOrNull
import org.json.JSONException
import org.json.JSONTokener
import kotlin.coroutines.resume

/** The rendered page content captured from a WebView. */
data class CapturedPage(val rawHtml: String?, val title: String?)

/**
 * Loads a URL in a WebView and returns the rendered DOM as
 * `document.documentElement.outerHTML` plus `document.title` — the same content the
 * browser extension and the iOS POC capture and send to `save-html`.
 *
 * Mirrors the iOS captor's resolution rules: extract after the first page-finished
 * (plus a short settle delay so script-rendered content is present), and on timeout
 * still extract whatever has rendered so far. A main-frame load failure yields null
 * fields instead — unlike iOS, Android's WebView renders its own error page into
 * the DOM, and extracting it would upload that error page as article content.
 * Never throws; the caller degrades to a URL-only save on null. All methods must
 * be called on the main thread (a WebView requirement).
 */
@SuppressLint("SetJavaScriptEnabled")
class HtmlCaptor(context: Context) {
	val webView: WebView = WebView(context).apply {
		settings.javaScriptEnabled = true
		settings.domStorageEnabled = true
		settings.userAgentString = AppConfig.WEB_VIEW_USER_AGENT
	}

	suspend fun capture(url: String, timeoutMillis: Long = 12_000, settleMillis: Long = 400): CapturedPage {
		val mainFrameLoaded = withTimeoutOrNull(timeoutMillis) {
			val loaded = awaitPageFinished(url)
			if (loaded) delay(settleMillis)
			loaded
		}
		if (mainFrameLoaded == false) {
			webView.stopLoading()
			return CapturedPage(rawHtml = null, title = null)
		}
		if (mainFrameLoaded == null) webView.stopLoading()
		val page = CapturedPage(
			rawHtml = evaluate("document.documentElement.outerHTML"),
			title = evaluate("document.title"),
		)
		webView.stopLoading()
		return page
	}

	/** Resumes true when the main frame finished, false when it failed to load. */
	private suspend fun awaitPageFinished(url: String): Boolean = suspendCancellableCoroutine { continuation ->
		webView.webViewClient = object : WebViewClient() {
			override fun onPageFinished(view: WebView, finishedUrl: String) {
				if (continuation.isActive) continuation.resume(true)
			}

			override fun onReceivedError(view: WebView, request: WebResourceRequest, error: WebResourceError) {
				if (request.isForMainFrame && continuation.isActive) continuation.resume(false)
			}
		}
		webView.loadUrl(url)
	}

	private suspend fun evaluate(js: String): String? = suspendCancellableCoroutine { continuation ->
		webView.evaluateJavascript(js) { raw -> continuation.resume(decodeJsString(raw)) }
	}

	/** `evaluateJavascript` returns a JSON-encoded value; unwrap a JSON string, else null. */
	private fun decodeJsString(raw: String?): String? {
		if (raw == null || raw == "null") return null
		return try {
			JSONTokener(raw).nextValue() as? String
		} catch (_: JSONException) {
			null
		}
	}
}

File: projects/native-apps/ios/Shared/HTMLCaptor.swift
import UIKit
import WebKit

/// The rendered page content captured from a WKWebView, plus the media type of
/// the loaded main-frame resource (`text/html` for a rendered page,
/// `application/pdf` for a PDF the captor declined to render, nil when the load
/// failed before a response). The media type lets the save journey pick the
/// matching `save-content` upload instead of degrading every non-HTML resource
/// to a URL-only crawl.
struct CapturedPage {
	let rawHtml: String?
	let title: String?
	let mediaType: String?
}

/// Loads a URL in an off-screen WKWebView and returns the rendered DOM as
/// `document.documentElement.outerHTML` plus `document.title` — the same
/// content the browser extension captures and uploads via `save-content`.
///
/// Resolves on first main-frame load completion (after a short settle delay so
/// script-rendered content is present) or when the timeout elapses, whichever
/// comes first. Never throws: a failed load yields a `CapturedPage` with nil
/// fields so the caller can degrade to a URL-only save.
@MainActor
final class HTMLCaptor: NSObject, WKNavigationDelegate {
	/// Host this (hidden) in a view so the web content reliably lays out and runs JS.
	let webView: WKWebView

	private var continuation: CheckedContinuation<CapturedPage, Never>?
	private var timeoutTask: Task<Void, Never>?
	private var settleSeconds: Double = 0.4
	private var detectedMediaType: String?

	override init() {
		let configuration = WKWebViewConfiguration()
		configuration.defaultWebpagePreferences.allowsContentJavaScript = true
		webView = WKWebView(
			frame: CGRect(x: 0, y: 0, width: 414, height: 896),
			configuration: configuration
		)
		super.init()
		webView.navigationDelegate = self
		webView.customUserAgent = AppConfig.webViewUserAgent
	}

	func capture(url: URL, timeout: TimeInterval) async -> CapturedPage {
		await withCheckedContinuation { continuation in
			self.continuation = continuation
			self.timeoutTask = Task { [weak self] in
				try? await Task.sleep(nanoseconds: UInt64(timeout * 1_000_000_000))
				await self?.finish(extractContent: true)
			}
			webView.load(URLRequest(url: url))
		}
	}

	// Records the main-frame resource's media type before deciding whether to
	// render it. A PDF is cancelled rather than loaded — a large PDF would blow
	// the share extension's memory budget — and finished as a non-extract capture
	// so the save journey fetches the bytes itself and uploads them as a file.
	// Setting `detectedMediaType` before `.cancel` (which provokes a
	// didFailProvisionalNavigation re-entry) means the media type is already
	// stamped no matter which path resolves `finish` first.
	nonisolated func webView(
		_ webView: WKWebView,
		decidePolicyFor navigationResponse: WKNavigationResponse,
		decisionHandler: @escaping (WKNavigationResponsePolicy) -> Void
	) {
		let decision = Self.navigationResponseDecision(
			mimeType: navigationResponse.response.mimeType,
			isMainFrame: navigationResponse.isForMainFrame
		)
		Task { @MainActor [weak self] in
			guard let self else { return decisionHandler(.allow) }
			switch decision {
			case .allow(let detectedMediaType):
				if let detectedMediaType { self.detectedMediaType = detectedMediaType }
				decisionHandler(.allow)
			case .captureAsFile(let mediaType):
				self.detectedMediaType = mediaType
				decisionHandler(.cancel)
				await self.finish(extractContent: false)
			}
		}
	}

	enum NavigationResponseDecision: Equatable {
		case allow(detectedMediaType: String?)
		case captureAsFile(mediaType: String)
	}

	nonisolated static func navigationResponseDecision(mimeType: String?, isMainFrame: Bool) -> NavigationResponseDecision {
		guard isMainFrame else { return .allow(detectedMediaType: nil) }
		if mimeType == "application/pdf" { return .captureAsFile(mediaType: "application/pdf") }
		return .allow(detectedMediaType: mimeType)
	}

	// WKNavigationDelegate's requirements are nonisolated; these hop to the main
	// actor (where WebKit already calls them) to touch the actor-isolated state.
	nonisolated func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
		Task { @MainActor [weak self] in
			guard let self else { return }
			try? await Task.sleep(nanoseconds: UInt64(self.settleSeconds * 1_000_000_000))
			await self.finish(extractContent: true)
		}
	}

	nonisolated func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
		Task { @MainActor [weak self] in await self?.finish(extractContent: true) }
	}

	nonisolated func webView(
		_ webView: WKWebView,
		didFailProvisionalNavigation navigation: WKNavigation!,
		withError error: Error
	) {
		Task { @MainActor [weak self] in await self?.finish(extractContent: false) }
	}

	private func finish(extractContent: Bool) async {
		guard let continuation else { return }
		self.continuation = nil
		timeoutTask?.cancel()
		timeoutTask = nil

		var page = CapturedPage(rawHtml: nil, title: nil, mediaType: detectedMediaType)
		if extractContent {
			let html = (try? await webView.evaluateJavaScript("document.documentElement.outerHTML")) as? String
			let title = (try? await webView.evaluateJavaScript("document.title")) as? String
			page = CapturedPage(rawHtml: html, title: title, mediaType: "text/html")
		}
		webView.stopLoading()
		continuation.resume(returning: page)
	}
}

extension HTMLCaptor: HTMLCapturing {
	/// The orchestrator-facing capture, pinned to the production timeout.
	func capture(url: URL) async -> CapturedPage { await capture(url: url, timeout: 12) }
}
